[yell.ow] 지하철 노선도 step4 리뷰 요청 부탁드립니다!

isangu opened the pull request:
안녕하세요 옐입니다.
기존 피드백 주셨던 사항을 반영하고, LineService가 LineDao, SectionService, StationService를 가지도록
SectionService를 생성하고 레이어를 분리해보았습니다.
많은 피드백 부탁드립니다!! 😆

kakao-yell-ow committed:
refactor : Line의 updateSectionWhenDeleted/Inserted 메서드 명을 addSection/deleteSection으로 변경 및 Sections의 getsize 삭제

kakao-yell-ow committed:
refactor : Service Layer 분리

- Line의 updateLineInfo, addSection, deleteSection을 updateLine 메서드 오버로딩으로 구현하도록 수정
- LineService가 LineDao, SectionService, StationService 를 가지도록 수정 및 Section 저장 로직 삭제
- SectionService 생성 및 구간 CRUD 기능 구현

Hue9010 commented on src/main/java/subway/station/StationResponse.java:
```diff
@@ -17,6 +17,10 @@ public StationResponse(Station station) {
         this.name = station.getName();
     }
 
+    public static StationResponse of(Station station) {
+        return new StationResponse(station.getId(), station.getName());
+    }
```
[이펙티브 자바 - 정적 팩토리 메소드 패턴](https://aerain.github.io/effective%20java/EJ-1/)
```java
1) from

매개 변수를 하나 받아서 해당 타입의 인스턴스를 반환하는 메서드에 주로 사용합니다.
Date date = Date.from(instnace);

2) of
여러 매개변수를 받아 적합한 타입의 인스턴스를 반환하는 집계 메서드에서 주로 사용합니다.
List list = List.of(1, 2, 3);
```
위의 명명 규칙은 제안 사항일 뿐이라 꼭 따를 필요는 없지만 많이 사용하는 규칙을 사용하는게 다른 작업자와의 협업에 도움이 됩니다.

Hue9010 commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,92 @@
+        sectionDao.deleteAllByLineId(lineId);
+    }
+
+    public List<StationResponse> getStationResponsesById(Long lineId, Long startStationId) {
+        List<StationResponse> responses = new ArrayList<>();
+        Sections sections = new Sections(sectionDao.findAllSectionsByLineId(lineId), startStationId);
+        for (Long stationId : sections.getStationsSortedSequence()) {
+            responses.add(new StationResponse(stationDao.findById(stationId)));
+        }
+        return responses;
+    }
```
더 이상 사용되지 않네요.

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -1,37 +1,35 @@
 
-    public static final String CONTAINS_BOTH_STATIONS_OR_NOT_EXISTS_STATIONS_ERROR_MESSAGE = "두 역이 모두 포함되어 있거나, 두 역 모두 포함되어 있지 않습니다.";
-    public static final String SECTIONS_SIZE_ERROR_MESSAGE = "구간이 하나이기 때문에 삭제할 수 없습니다.";
     private LineDao lineDao;
```
dao의 수정은 없어서 여기다가 달게요. 🙏 

```java
@Repository
public class LineDao {
    public void deleteById(Long id) {
        int deletedRow = jdbcTemplate.update(DELETE_LINE_BY_ID, id);
        if(deletedRow < NO_DELETED_ROW) {
            throw new InvalidLineArgumentException(NO_MATCHING_LINE_ERROR_MESSAGE);
        }
    }
```

사실 이전 단계까지의 편의성을 위해 보류해두고 있던 사항인데 dao는 Data Access Object로 간단히 말하면 db와의 접근을 위한 객체로 말그대로 DB에 접근하는 역할 외 로직을 넣지 않습니다. deleteById로 삭제를 하는게 역할의 끝입니다. Line 68, 69에서의 실제 삭제되었는지는 service layer에서 확인해줘야 합니다.

findById 반환 타입에 Optinal을 적용해보세요. Optinal을 사용해보는게 목적이니 모두 다하기보다는 LineDao 한정으로 적용해봐도 충분해 보여요. 😉

Hue9010 commented on src/main/java/subway/section/SectionService.java:
```diff
@@ -0,0 +1,92 @@
+import java.util.List;
+
+@Service
+public class SectionService {
```
👍

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -44,86 +42,59 @@ public Line findById(Long id) {
+        line.updateLine(newStartEndSection);
         lineDao.updateById(line);
-        sectionDao.save(newSection);
+        sectionService.save(newSection);
```
line 79에서 라인은 정상적으로 갱신하고 line 80에서 section 저장시 에러가 발생하면 데이터 상 문제가 발생 할 것 같은데 이에 대해서는 무엇을 고려 해야할까요? 
힌트를 드리자면 DB 쪽에서 자주 사용하는 용어가 있습니다. 😉 (DB에 국한된 용어가 아니란 점 유의해주세요.)

Hue9010 reviewed (changes requested):
이전부터 layer구분을 해주셔서 빠르게 반영해주셨네요. 👍
몇가지 개선하면 좋을 사항 남겨뒀는데 궁금한 부분 질문 주세요!

kakao-yell-ow committed:
refactor : LineDao 리팩터링 및 사용하지 않는 변수, 메서드 리팩터링

- LineDao의 Error Handling을 LineService 레이어로 이동
- SectionService의 사용하지 않는 getStationResponsesById 메서드 삭제
- StationResponse의 정적 팩토리 메서드 네임 of -> from으로 변경

kakao-yell-ow committed:
refactor : Section, Station 리팩터링

- SectionDao 의 ErrorHandling 을 SectionService 로 이동
- StationDao 의 ErrorHandling 을 StationService 로 이동

idealgu commented:
좋은 피드백 감사드립니다!! 😄😄
말씀해주셨던 부분 반영하여 Dao의 Error 핸들링은 각 Service 레이어가 처리하도록 변경하였고
DB 정합성을 위해 여러 DB의 정보가 변경되는 부분은 @Transactional 어노테이션을 붙여 처리했습니다!

Hue9010 commented on src/main/java/subway/line/LineService.java:
```diff
@@ -1,129 +1,113 @@
+        this.stationService = stationService;
     }
 
+    @Transactional
```
👍 
Transactional에 대한 기초적인 설명을 드리려고했는데 subway-service 강의 내용에 포함되어 있군요. 😉
참고로 readOnly에 대해서도 사용하는게 좋습니다. https://www.inflearn.com/questions/7185

Hue9010 reviewed (approved):
깔끔히 잘 적용해주셨네요! 👍 
이번 단계는 여기서 머지하도록 할게요. 고생하셨습니다!!
단계는 끝났지만 궁금하거나 질문하고 싶은 사항 있으면 언제든지 문의주세요!! 😉 
다음 단계도 화이팅입니다!!
